addons/sale_coupon/models/sale_order.py

```diff
@@ -363,9 +368,11 @@ def _remove_invalid_reward_lines(self):
         self.ensure_one()
         order = self
 
-        applicable_programs = order._get_applicable_no_code_promo_program() + order._get_applicable_programs() + order._get_valid_applied_coupon_program()
-        applicable_programs = applicable_programs._keep_only_most_interesting_auto_applied_global_discount_program()
         applied_programs = order._get_applied_programs_with_rewards_on_current_order() + order._get_applied_programs_with_rewards_on_next_order()
+        applicable_programs = self.env['sale.coupon.program']
+        if applied_programs:
+            applicable_programs = order._get_applicable_no_code_promo_program(applied_programs.ids) + order._get_applicable_programs(applied_programs.ids) + order._get_valid_applied_coupon_program()
+            applicable_programs = applicable_programs._keep_only_most_interesting_auto_applied_global_discount_program()
```

Comment:
the `ids_in` logic indeed breaks the current tested behavior, it should probably be avoided.

Reply:
this part of the code is so logically strange in my opinion.
Why would we need to look for applicable programs in this part of the code ? I can understand on the part where we want to add new coupons, but here it is to delete existing one that should not be apply.
The reasoning behind it is strange in my opinion. If it is necessary, it should at least be store somewhere the first time it is computed to avoid recalculation as it is quite heavy.

Reply:
I agree the code logic is strange (and difficult to understand), but the problem we are facing and trying to resolve is performance, there shouldn't be any need to modify the functional behavior (and the tests, because you're breaking some of them)...

Unless we really have a "functional" bugfix ticket, this shouldn't be modified in stable IMO ...